### ISSUE TITLE
Differentiability

### ISSUE THREAD

**Issue by andreped** [Surayuth/DifferentiableMacenko#2]
Are you sure your code is differential? I would think this if statement would say otherwise.
https://github.com/Surayuth/DifferentiableMacenko/blob/fa77ea0335316f4cb832f2fdbd74c31d89572134/main.py#L64

See line 44 for how I handled it:
https://github.com/andreped/torchstain/commit/ef59f582db5aee6b4d9ed0bdf062fc04458d4e0b

In order to use the GPU you cannot have to use the torch.where instead of pythonic if statements.

But have you defined some neural network? What is the goal of your task? On which data? Why did you want to use the H channel in the loss? Is this part of your studies/work/msc/phd-stuff?

**Comment by Surayuth** [comment]
Thanks for looking into my code.

I try to move all variables to GPU, see  DifferentiableMacenko/use_gpu.py
at main · Surayuth/DifferentiableMacenko (github.com)
<https://github.com/Surayuth/DifferentiableMacenko/blob/main/use_gpu.py>,
and call backward(). The code still runs fine. Is there anything wrong with
it?

I'm working on my MSC right now (it's really hard so far to find a new idea
to work on). The main idea of this work is shown below

IHC image ---> model_1(normalize images to target dataset)    ---> output 1
---> model_2(any network) --> stain concentration

The fact that I pass the output from model 1 further is because I want to
ensure that the structure of the stain in the original images is conserved
which is why I want the model_2 (which can be Macenko's) to be
differentiable.

**Comment by Surayuth** [comment]
Using percentile to segment the stained region would be simpler in model_2
(instead of using the whole Marcenko process) but choosing the right
percentile for all datasets is not possible.

**Comment by andreped** [comment]
But is the idea to use the extracted H image from Macenko as the golden standard to train a neural network that does the exact same thing?

But if you use Macenko as model 1, **it does not matter if it is differentiable**. You do not update any weights there. model_1 would be static. You only care about updating the weights of model_2.

Also, by stain concentration, what do you mean? Do you wish to segment the image containing the H stain, or do you wish to directly predict the amount of H that is present in an image?

**Comment by Surayuth** [comment]
I want the H stain segmented from model_2 to be as similar as possible to the original H stain segmented from the input of model_1. This problem arises from the fact that the network I use to normalize images, model_1, doesn't preserve the amount of H stain sometimes. For example, the H stain of the output of model_1 looks fade. 

Now, if the model_1 is really good at normalizing images, it should preserve the structure of the stained area (the output from model_2) as well. Thus, I will compare the H stain, generated from model_2, with the original H stain extracted from the input of model_1.  The result is a loss which I can use to improve model_1 to better at preserving the structure of the stain H.

Existing work will use a pre-trained model to extract stain in model_2 but this requires a gt label. So I want to find another way to do the job of model_2 without requiring the gt label.

**Comment by andreped** [comment]
Oh, okay. I understand. Interesting. How were you thinking of training the neural network then? How would you define the loss? Did you find any related work on this perhaps?

But yes, if you are able to define some suitable objective function, you could train the model_2 somewhat unsupervised.

**Comment by Surayuth** [comment]
MSEloss per pixel should be fine as I just want the stain of the normalized image to appear as the stain segmented from the input of model_1.  
The problem I found from using Macenko's method to extract the stain is 
1. It's slow. I think this comes from the fact that it has to do SVD and solve leastsq problem.
2.  It's not stable, i.e, SVD is not converging for some images.

Most related works found so far use semantic guidance or saliency map.

**Comment by andreped** [comment]
As I said, there is no need to use Macenko end-to-end during training. You can just estimate the H image in preprocessing and train using the H&E-stained image as input and use the H image in the loss stuff, if you want to. Hence, not using Macenko end-to-end would speed up training.

Could you reference some articles that talk about this, such that I understand why this is a good idea.

And you model_2 which architecture are you using? Just a simple autoencoder that takes the pixels as input, or are you using a convolutional autoencoder?

**Comment by Surayuth** [comment]
> Could you reference some articles that talk about this, such that I understand why this is a good idea.

There isn't an article that talks about this. I just remember this algorithm and think that maybe I can use it to train my networks unsupervised.

> And you model_2 which architecture are you using? Just a simple autoencoder that takes the pixels as input, or are you using a convolutional autoencoder?

Right now, I haven't picked any architecture yet. A simple autoencoder or CNN that can roughly segment the stain would be good. But if model_2 can be trained unsupervised too, it would be great.

**Comment by andreped** [comment]
Why dont you just perform pixel-wise clustering to extract the H image
after normalization? That is quite easy to do, and does not necessarily
require any deep learning. I would try a traditional clustering pipeline
first to serve as baseline.

**Comment by Surayuth** [comment]
Thanks for the idea. Can you provide any material or related paper that I can follow?

**Comment by andreped** [comment]
Well, that was just an idea.

But why do you want to estimate the H image? What are you going to use it for? Is the goal simply just to improve stain normalization?

**Comment by Surayuth** [comment]
> Is the goal simply just to improve stain normalization?

Yes, that's the goal. The amount of H is not my main concern.

**Comment by andreped** [comment]
Then I would much rather start by making a simple autoencoder or similar and setup the training scheme for that one, before looking into how to optimize such a baseline by introducing the H image from Macenko.

I would start doing making an autoencoder similar to what they did in this study:
https://arxiv.org/pdf/1902.06543.pdf

See Figure 4 for how it was trained. Remember to have data from a specific scanner/lab as reference when training.
I would start by making a simple autoencoder which just takes the pixels as input (no convolutions). 
I believe that should work well on this task.

Then of course, you could see if you could improve this baseline, by introducing the H image in the loss somehow.

Stain normalization is a very interesting topic. I am not sure using NNs for this is the best (in terms of generalizability), but it is a very popular research topic (using NNs for this task).

Maybe you could let me know when you have setup up the autoencoder? 

Also, it might be smart to make this repository private. To avoid anyone stealing your/our ideas, if they becomes publishworthy in the future.
